tests/unit/json_test.c

```diff
@@ -1908,116 +1909,92 @@ static void test_json_null_not_null(void)
-        const JsonElement *const b_child = JsonObjectGet(b, key);
-        if (b_child == NULL || !compare_json_object_merge_deep(a_child, b_child))
-        {
-            return false;
-        }
-    }
-    return true;
-}
-
-static bool compare_json_object_merge_deep(const JsonElement *const a, const JsonElement *const b)
-{
-    const JsonType type = JsonGetType(a);
-    if (type != JsonGetType(b))
-    {
-        return false;
-    }
-
-    switch (type)
-    {
-    case JSON_TYPE_OBJECT:
-        return compare_object_json_object_merge_deep(a, b);
-    case JSON_TYPE_ARRAY:
-        return compare_array_json_object_merge_deep(a, b);
-    default:
-        return JsonCompare(a, b) == 0;
-    }
-}
-
 static bool check_json_object_merge_deep(const char *base_raw, const char *extra_raw, const char *expected_raw)
 {
+    LogSetGlobalLevel(LOG_LEVEL_DEBUG);
+    bool pass = true;
+    JsonElement *actual = NULL;
     JsonElement *base = NULL;
+    JsonElement *extra = NULL;
+    JsonElement *expected = NULL;
+    char *expected_string = NULL;
+    char *actual_string = NULL;
+
     if (JsonParse(&base_raw, &base) != JSON_PARSE_OK)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *extra = NULL;
     if (JsonParse(&extra_raw, &extra) != JSON_PARSE_OK)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *expected = NULL;
     if (JsonParse(&expected_raw, &expected) != JSON_PARSE_OK) {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonElement *actual = JsonObjectMergeDeep(base, extra);
+    expected_string = JsonToString(expected);
+
+    actual = JsonObjectMergeDeep(base, extra);
     if (actual == NULL || actual == base)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    /* JsonCompare is naive and does not handle containers properly, thus we
-     * need a custom comparison function to test for equality. */
-    if (!compare_json_object_merge_deep(actual, expected))
+    actual_string = JsonToString(actual);
+
+    if (JsonCompare(actual, expected))
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
     JsonDestroy(actual);
+    actual = NULL;
     actual = JsonObjectMergeDeepInplace(base, extra);
+
+    // JsonObjectMergeDeepInplace() returns the first parameter (base) so actual should point to base
     if (actual != base)
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    if (!compare_json_object_merge_deep(actual, expected))
+    if (JsonCompare(actual, expected))
     {
-        return false;
+        pass = false;
+        goto finish;
     }
 
-    JsonDestroy(base);
-    JsonDestroy(extra);
-    JsonDestroy(expected);
+    finish:
+        if (actual_string)
+        {
+            if (pass == false)
+                printf("merged json  : %s\n", actual_string);
```

Comment:
> *Always* use curly braces
   - Ref [CONTRIBUTING.md](https://github.com/cfengine/core/blob/master/CONTRIBUTING.md#readability--maintainability)

Reply:
I suspected as much as I wrote it but couldn't help myself ;)